presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java

```diff
@@ -0,0 +1,201 @@
+    private final Closer closer = Closer.create();
+    private final SingleStreamSpillerFactory spillerFactory;
+    private final SpillContext spillContext;
+    private final SynchronizedAggregatedMemoryContext memoryContext;
+
+    private final PageBuilder[] pageBuilders;
+    private final Optional<SingleStreamSpiller>[] spillers;
+
+    private boolean readingStarted;
+    private Set<Integer> spilledPartitions = new HashSet<>();
+
+    public GenericPartitioningSpiller(
+            List<Type> types,
+            PartitionFunction partitionFunction,
+            SpillContext spillContext,
+            AggregatedMemoryContext memoryContext,
+            SingleStreamSpillerFactory spillerFactory)
+    {
+        requireNonNull(spillContext, "spillContext is null");
+
+        this.types = requireNonNull(types, "types is null");
+        this.partitionFunction = requireNonNull(partitionFunction, "partitionFunction is null");
+        this.spillerFactory = requireNonNull(spillerFactory, "spillerFactory is null");
+        this.spillContext = closer.register(requireNonNull(spillContext, "spillContext is null"));
+
+        requireNonNull(memoryContext, "memoryContext is null");
+        closer.register(memoryContext::close);
+        this.memoryContext = synchronizedMemoryContext(memoryContext);
+
+        int partitionCount = partitionFunction.getPartitionCount();
+        this.pageBuilders = new PageBuilder[partitionCount];
+        //noinspection unchecked
+        this.spillers = new Optional[partitionCount];
+
+        for (int partition = 0; partition < partitionCount; partition++) {
+            pageBuilders[partition] = new PageBuilder(types);
+            spillers[partition] = Optional.empty();
+        }
+    }
+
+    @Override
+    public synchronized Iterator<Page> getSpilledPages(int partition)
+    {
+        readingStarted = true;
+        getFutureValue(flush(partition));
+        spilledPartitions.remove(partition);
+        return getSpiller(partition).getSpilledPages();
+    }
+
+    @Override
+    public synchronized void verifyAllPartitionsRead()
+    {
+        verify(spilledPartitions.isEmpty(), "Some partitions were spilled but not read: %s", spilledPartitions);
+    }
+
+    @Override
+    public synchronized PartitioningSpillResult partitionAndSpill(Page page, IntPredicate spillPartitionMask)
+    {
+        requireNonNull(page, "page is null");
+        requireNonNull(spillPartitionMask, "spillPartitionMask is null");
+        checkArgument(page.getChannelCount() == types.size(), "Wrong page channel count, expected %s but got %s", types.size(), page.getChannelCount());
+
+        checkState(!readingStarted, "reading already started");
+        IntArrayList unspilledPositions = partitionPage(page, spillPartitionMask);
+        ListenableFuture<?> future = flushFullBuilders();
+
+        return new PartitioningSpillResult(future, mask(page, unspilledPositions.toIntArray()));
```

Comment:
why don't we use something like `page.mask(unspilledPositions.toIntArray())` as the signature?